src/elf.rs

```diff
@@ -814,14 +760,19 @@ impl<'a> Elf<'a> {
 
         // alignment required for below
         let mut symtab_offset = strtab_offset + self.sizeof_strtab as u64;
-        let sizeof_symbol = Symbol::size(self.ctx.container) as u64;
-        Self::align(&mut symtab_offset, sizeof_symbol);
-        let mut reloc_offset = symtab_offset + sizeof_symtab as u64;
-        let sizeof_relocation = Relocation::size(true, self.ctx) as u64;
-        Self::align(&mut reloc_offset, sizeof_relocation);
+        let symtab_align = self.ctx.size() as u64;
+        Self::align(&mut symtab_offset, symtab_align);
+        let mut symtab_shndx_offset = symtab_offset + sizeof_symtab as u64;
+        let symtab_shndx_align = 4;
+        if sizeof_symtab_shndx > 0 {
+            Self::align(&mut symtab_shndx_offset, symtab_shndx_align);
+        }
+        let mut reloc_offset = symtab_shndx_offset + sizeof_symtab_shndx;
+        let reloc_align = self.ctx.size() as u64;
```

Comment:
I'm confused, how could this be the same as `Relocation::size(true, self.ctx)`? Also the naming is somewhat inconsistent; the function calls it `sizeof_t`, but it's called align here.

Reply:
It's not the same, that's why I changed it. The structures aren't aligned to their size, they're aligned to the largest alignment required by their fields, which is an address, and `self.ctx.size()` gives the size of an address.

Reply:
Also, this matches the alignment that is set for the relocation sections, etc: https://github.com/m4b/faerie/blob/49aac0ca5b49c6ef95a57a95d8a006ab73e998e5/src/elf.rs#L304 (well, that should be using ctx.size() too, but hopefully you see what I mean).

Reply:
Yes you're totally right, sorry for the confusion - very good catch! Probably the math turns out to yield a an address which is also aligned to a pointer width, but we should be correct, thank you!

Just a note: I didn't think the alignment is aligned to the largest alignment required by their fields, as far as I understood, it's the alignment required to make arbitrary acceses safe w.r.t. an aligned pointer, e.g., this formula must be true of for a correctly aligned `address`:
```
(address + (sizeof_t * i)) mod sizeof_ptr = 0
```
e.g., when our starting address is aligned, offsetting by index `i * sizeof foo` is guaranteed to yield an aligned pointer:
```
address: [ foo; foo; foo; ]
  aligned ^
                       ^ (address + sizeof_foo) mod sizeof_ptr = 0
```
Maybe it's the same thing to say aligned to largest alignment required by fields, but I've always understood it as above, but perhaps that's wrong?

Reply:
> it's the alignment required to make arbitrary acceses safe w.r.t. an aligned pointer

Yes that's true, but generally the way you do it that is if `align` is the alignment required for a field within a structure, then you layout the structure so that the field is placed at an `offset` within the structure which satisfies `offset % align == 0`. Given that, it is then enough to ensure that the structure is placed at an address such that `address % align == 0` for every field's alignment, which is equivalent to ensuring that the address uses the largest alignment required by fields.

The other thing that is done is that padding is added to the end of the structure to ensure that `size % align == 0`. This allows arrays of the structure to work automatically.